### PR TITLE
Make data relocation at the end of workflow  runs faster

### PR DESCRIPTION
**Pull request by psafont**
- [x] When moving files, only calculate realpaths of links instead of doing it
for all the files in the destination directory
- [x] Avoid stating through all the files in the destination folder
- [ ] Avoid stating through all the folders in the destination folder

### CODE REVIEW

File: cwltool/process.py [hunk 1]
```diff
@@ -616,9 +617,10 @@ def _init_job(self, joborder, runtimeContext):
             tmpdir = runtimeContext.docker_tmpdir or "/tmp"
             stagedir = runtimeContext.docker_stagedir or "/var/lib/cwl"
         else:
-            outdir = fs_access.realpath(runtimeContext.outdir or
-                tempfile.mkdtemp(prefix=getdefault(runtimeContext.tmp_outdir_prefix,
-                                                   DEFAULT_TMP_PREFIX)))
+            outdir = (runtimeContext.outdir or
+                      tempfile.mkdtemp(prefix=getdefault(runtimeContext.tmp_outdir_prefix,
+                                                         DEFAULT_TMP_PREFIX)))
+            outdir = fs_access.realpath(outdir)
```
Comment (mr-c) [inline cwltool/process.py]: Why? I wouldn't call `fs_access.realpath` on the output of `tempfile.mkdtemp`..

Comment (psafont) [reply cwltool/process.py]: I did not mean to change the behaviour here, just the formatting. having said that, I understood that `fs_access.realpath` was already being called on the output of `tempfile.mkdtemp` Did I misunderstand the statement?

Comment (mr-c) [reply cwltool/process.py]: Sorry, missed that this was a pure reformat. I'll investigate further, elsewhere.

**Comment by psafont** [comment]
I'm happy with the performance of the code right now, although it is still scanning the destination folder. I've tried to simplify the code so further changes are easier to do, if they are needed. (like saving all symlinks in the outputs and then only changing those, without going through all the folder in the output folder)

The only thing I'm not sure about is if I included the `scandir` in the dependencies correctly since it's not needed for python 3.5+

I've not replaced the `os.walk` in pathmapper since I think that adding code for a library to make it a bit faster does not seem like it's going to give a tangible performance improvement.

**Comment by mr-c** [comment]
Thank you @psafont for your work on this.

Did you see the conformance test failures?

https://ci.commonwl.org/job/cwltool-pr-conformance-multiver/1718/version=v1.0/testReport/junit/conformance_test_v1/0/Test_directory_output/
https://ci.commonwl.org/job/cwltool-pr-conformance-multiver/1718/version=v1.0/testReport/junit/conformance_test_v1/0/Test_directory_literal_output_created_by_ExpressionTool/

File: cwltool/process.py [hunk 2]
```diff
@@ -287,20 +279,38 @@ def relocateOutputs(outputObj,             # type: Union[Dict[Text, Any],List[Di
     if action not in ("move", "copy"):
         return outputObj
 
-    def moveIt(src, dst):
+    def _collectDirEntries(obj):
+        # type: (Union[Dict[Text, Any], List[Dict[Text, Any]]], List[Dict[Text, Any]]) -> Iterator[Dict[Text, Any]]
```
Comment (mr-c) [inline cwltool/process.py]: `error: Type signature has too many arguments`

Comment (psafont) [reply cwltool/process.py]: Fixed! Tried to also fix the type problems from the consolidation into a single temp variable, then processing it contents that, but couldn't make it work.

**Comment by psafont** [comment]
The two failures might have been because I used the wrong variable in a specific location. (strange the normal tests do not complain)

**Comment by mr-c** [comment]
> Not sure why this happens in Travis' python 3.5 when it fine on python 3.6+
We run mypy to do the type checking using Python 3.5 (in both py3 and py2 mode) as the results are presumed to be the same from any version. Perhaps we should check the types in each of py{3.5,6,7}?

The error
```
cwltool/process.py:13: error: Module 'os' has no attribute 'scandir'
cwltool/process.py:15: error: Cannot find module named 'scandir'
```
Is due to lack of type annotations for scandir.

**Comment by psafont** [comment]
Getting back to the complex flow of an inlined function makes it work. I tried to make the behaviour clearer by removing an indent level.

It's fishy, though, if the order is to 'move', why doesn't move actually move the files and they are copied instead?

**Comment by psafont** [comment]
Seeing the coverage report, the most complex parts of the function are not being used: Mainly the folder merging on the relocate subfunction and relinking of symlinks at the end of the process.

Is there any case where this functionality may be useful?
If it is, it may be worth adding it to conformance tests since it seems to be a rare corner case.
If it isn't it may be worth rescoping the purpose of the function and surrounding code to simplify it.

**Comment by tetron** [comment]
@psafont could you add some unit tests?  These functions are pretty self contained so it should be straightforward to write unit tests for them.  The solution to not enough coverage for the tricky bits is to improve the coverage :-)

**Comment by psafont** [comment]
Since this seems to be part of an extension I'll add tests to `tests/test_ext.py`
